src/firestore/delete.ts

```diff
@@ -0,0 +1,486 @@
+import { FirebaseError } from "../error";
+import * as logger from "../logger";
+import * as utils from "../utils";
+
+// Datastore allowed numeric IDs where Firestore only allows strings. Numeric IDs are
+// exposed to Firestore as __idNUM__, so this is the lowest possible negative numeric
+// value expressed in that format.
+const MIN_ID = "__id-9223372036854775808__";
+
+export class FirestoreDelete {
+  /**
+   * Progress bar shared by the class.
+   */
+  static progressBar = new ProgressBar("Deleted :current docs (:rate docs/s)\n", {
+    total: Number.MAX_SAFE_INTEGER,
+  });
+
+  public isDocumentPath: boolean;
+  public isCollectionPath: boolean;
+  public path: string;
+
+  private project: string;
+  private recursive: boolean;
+  private shallow: boolean;
+  private allCollections: boolean;
+
+  private allDescendants: boolean;
+  private root: string;
+  private parent: string;
+
+  /**
+   * Construct a new Firestore delete operation.
+   *
+   * @constructor
+   * @param {string} project the Firestore project ID.
+   * @param {string | undefined} path path to a document or collection.
+   * @param {boolean} options.recursive true if the delete should be recursive.
+   * @param {boolean} options.shallow true if the delete should be shallow (non-recursive).
+   * @param {boolean} options.allCollections true if the delete should universally remove all collections and docs.
+   */
+  constructor(
+    project: string,
+    path: string | undefined,
+    options: { recursive?: boolean; shallow?: boolean; allCollections?: boolean }
+  ) {
+    this.project = project;
+    this.path = path || "";
+    this.recursive = Boolean(options.recursive);
+    this.shallow = Boolean(options.shallow);
+    this.allCollections = Boolean(options.allCollections);
+
+    // Remove any leading or trailing slashes from the path
+    this.path = this.path.replace(/(^\/+|\/+$)/g, "");
+
+    this.allDescendants = this.recursive;
+    this.root = "projects/" + project + "/databases/(default)/documents";
+
+    const segments = this.path.split("/");
+    this.isDocumentPath = segments.length % 2 === 0;
+    this.isCollectionPath = !this.isDocumentPath;
+
+    // this.parent is the closest ancestor document to the location we're deleting.
+    // If we are deleting a document, this.parent is the path of that document.
+    // If we are deleting a collection, this.parent is the path of the document
+    // containing that collection (or the database root, if it is a root collection).
+    this.parent = this.root;
+    if (this.isCollectionPath) {
+      segments.pop();
+    }
+    if (segments.length > 0) {
+      this.parent += "/" + segments.join("/");
+    }
+
+    // When --all-collections is passed any other flags or arguments are ignored
+    if (!options.allCollections) {
+      this._validateOptions();
+    }
+  }
+
+  /**
+   * Validate all options, throwing an exception for any fatal errors.
+   */
+  private _validateOptions() {
```

Comment:
Private functions don't need to be prefixed with underscores. Applies throughout.